### PR TITLE
feat(dspy): add vector, hybrid and fulltext search support in azure ai search module

### PR DESCRIPTION
**Pull request by HARISHKUMAR1112001**
Feature Addition: Updating AzureAISearchRM Class

This pull request aims to enhance the functionality of the `AzureAISearchRM` class by introducing support for additional search capabilities. Specifically, the class is updated to enable `Vector Search`, `Hybrid Search`, and `Full Text Search` with `Semantic ReRanker`, providing users with more versatile and powerful search options.

Changes Included:

- Integration of `Vector Search`, `Hybrid Search`, and `Full Text Search` functionalities.
- Implementation of `Semantic ReRanker` to enhance search results with semantic understanding.
- Addition of necessary methods and attributes to support the new features.
- Updates to the documentation to reflect the expanded capabilities of the class.

Checks:

 Pre-Commit checks are passing (locally and remotely)
 Title of your PR / MR corresponds to the required format
 Commit message follows required format {label}(dspy): {message}

If there are any further improvements or adjustments required, please feel free to provide feedback. This pull request aims to enhance the functionality and usability of the `AzureAISearchRM` class, and any suggestions for refinement are highly appreciated.

### CODE REVIEW

**Comment by HARISHKUMAR1112001** [comment]
@insop @arnavsinghvi11 , Can I get the review on this PR.

File: dspy/retrieve/azureaisearch_rm.py [hunk 1]
```diff
@@ -89,83 +153,161 @@ def __init__(
...
         semantic_ranker: bool = False,
         filter: str = None,
         query_language: str = "en-Us",
         query_speller: str = "lexicon",
         use_semantic_captions: bool = False,
         query_type: Optional[QueryType] = QueryType.FULL,
         semantic_configuration_name: str = None,
-
+        is_vector_search: Optional[bool] = False,
+        is_hybride_search: Optional[bool] = False,
```
Comment (arnavsinghvi11) [inline dspy/retrieve/azureaisearch_rm.py:169]: can you refactor all mentions of `is_hybride_search` to be `is_hybrid_search`

File: dspy/retrieve/azureaisearch_rm.py [hunk 2]
```diff
@@ -68,13 +107,38 @@ class AzureAISearchRM(dspy.Retrieve):
...
 
+        get_embeddings(
+            self,
+            query: str,
+            k_nearest_neighbors: int,
+            field_vector: str
+        ) -> List | Any:
+            Returns embeddings for the given query.
+
+        check_sementic_configuration(
```
Comment (arnavsinghvi11) [inline dspy/retrieve/azureaisearch_rm.py:135]: spelling error - can you refactor all uses of `sementic` to be `semantic`?

File: dspy/retrieve/azureaisearch_rm.py [hunk 3]
```diff
@@ -89,83 +153,161 @@ def __init__(
...
+                    semantic_configuration_name=semantic_configuration_name,
+                    top=top,
+                )
+            else:
+                results = client.search(
+                    search_text=None,
+                    filter=filter,
+                    vector_queries=[vector_query],
+                    vector_filter_mode=vector_filter_mode,
+                    top=top,
```
Comment (arnavsinghvi11) [inline dspy/retrieve/azureaisearch_rm.py:243]: not familiar with the semantics here myself but does this require a `query_caption` parameter like the other conditions?

File: dspy/retrieve/azureaisearch_rm.py [hunk 4]
```diff
@@ -177,25 +319,77 @@ def forward(self, query_or_queries: Union[str, List[str]],  k: Optional[int]) ->
...
+                self.azure_openai_client.embeddings.create(input=query, model=self.openai_embed_model).data[0].embedding
+            )
+            vector_query = VectorizedQuery(
+                vector=embedding, k_nearest_neighbors=k_nearest_neighbors, fields=field_vector,
+            )
+            return [vector_query]
+        else:
+            return self.embedding_func(query)
+
+    def check_sementic_configuration(self, semantic_configuration_name, query_type):
```
Comment (arnavsinghvi11) [inline dspy/retrieve/azureaisearch_rm.py:382]: refactor function name for spelling

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Thanks @HARISHKUMAR1112001 ! left a few minor comments. good to merge after that!

File: dspy/retrieve/azureaisearch_rm.py (hunk 3 above)
Comment (HARISHKUMAR1112001) [reply dspy/retrieve/azureaisearch_rm.py:243]: Yes, but that is optional parameter and I have provided the setting for the same if you are using semantic reranker. Just missed.

**Comment by HARISHKUMAR1112001** [comment]
Thanks @arnavsinghvi11 for the quick review. I have updated the code with correct spellings.

**Comment by arnavsinghvi11** [comment]
Thanks @HARISHKUMAR1112001 !

**Comment by SimplyJuanjo** [comment]
@HARISHKUMAR1112001 @arnavsinghvi11 i'm getting the following error with the following code:

requirements.txt
git+https://github.com/stanfordnlp/dspy.git
azure-search-documents==11.6.0b1

`
client = openai.AzureOpenAI(
        base_url=f"{openai.api_base}/openai/deployments/{deployment_id}/extensions",
            api_key=openai.api_key,
            api_version=openai.api_version
        )

        azure_search = AzureAISearchRM(
        search_service_name=search_endpoint,
        search_api_key=search_key,
        search_index_name=search_index_name,
        field_text="content",
        k=3,
        azure_openai_client=client,
        openai_embed_model="text-embedding-ada-002",
        )

        dspy.settings.configure(rm=azure_search)
        retrieve = dspy.Retrieve(k=3)
        retrieval_response = retrieve(name).passages

        response=''
        for result in retrieval_response:
            print("Text:", result, "\n")
            response += result + "\n"


error:


[2024-04-18T08:45:17.286Z] Executed 'Functions.f29bot2' (Failed, Id=60fe1453-0587-4435-bd53-525ca750323c, Duration=222ms)
[2024-04-18T08:45:17.286Z] System.Private.CoreLib: Exception while executing function: Functions.f29bot2. System.Private.CoreLib: Result: Failure
[2024-04-18T08:45:17.286Z] Exception: ServiceRequestError: <urllib3.connection.HTTPSConnection object at 0x7138005e6140>: Failed to resolve 'https' ([Errno -2] Name or service not known)
[2024-04-18T08:45:17.286Z] Stack:   File "/usr/lib/azure-functions-core-tools-4/workers/python/3.10/LINUX/X64/azure_functions_worker/dispatcher.py", line 495, in _handle__invocation_request
[2024-04-18T08:45:17.286Z]     call_result = await self._loop.run_in_executor(
[2024-04-18T08:45:17.286Z]   File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
[2024-04-18T08:45:17.286Z]     result = self.fn(*self.args, **self.kwargs)
[2024-04-18T08:45:17.286Z]   File "/usr/lib/azure-functions-core-tools-4/workers/python/3.10/LINUX/X64/azure_functions_worker/dispatcher.py", line 768, in _run_sync_func
[2024-04-18T08:45:17.286Z]     return ExtensionManager.get_sync_invocation_wrapper(context,
[2024-04-18T08:45:17.286Z]   File "/usr/lib/azure-functions-core-tools-4/workers/python/3.10/LINUX/X64/azure_functions_worker/extension.py", line 215, in _raw_invocation_wrapper
[2024-04-18T08:45:17.286Z]     result = function(**args)
[2024-04-18T08:45:17.286Z]   File "/home/juanjo/juanjo_repos/functions/f29bot2/__init__.py", line 54, in main
[2024-04-18T08:45:17.286Z]     retrieval_response = retrieve(name).passages
[2024-04-18T08:45:17.286Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dspy/retrieve/retrieve.py", line 30, in __call__
[2024-04-18T08:45:17.286Z]     return self.forward(*args, **kwargs)
[2024-04-18T08:45:17.286Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dspy/retrieve/retrieve.py", line 39, in forward
[2024-04-18T08:45:17.286Z]     passages = dsp.retrieveEnsemble(queries, k=k,**kwargs)
[2024-04-18T08:45:17.286Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dsp/primitives/search.py", line 57, in retrieveEnsemble
[2024-04-18T08:45:17.286Z]     return retrieve(queries[0], k, **kwargs)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dsp/primitives/search.py", line 12, in retrieve
[2024-04-18T08:45:17.287Z]     passages = dsp.settings.rm(query, k=k, **kwargs)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dspy/retrieve/retrieve.py", line 30, in __call__
[2024-04-18T08:45:17.287Z]     return self.forward(*args, **kwargs)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dspy/retrieve/azureaisearch_rm.py", line 328, in forward
[2024-04-18T08:45:17.287Z]     results = self.azure_search_request(
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/dspy/retrieve/azureaisearch_rm.py", line 288, in azure_search_request
[2024-04-18T08:45:17.287Z]     sorted_results = sorted(results, key=lambda x: x["@search.score"], reverse=True)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/search/documents/_paging.py", line 54, in __next__
[2024-04-18T08:45:17.287Z]     return next(self._page_iterator)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/paging.py", line 75, in __next__
[2024-04-18T08:45:17.287Z]     self._response = self._get_next(self.continuation_token)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/search/documents/_paging.py", line 124, in _get_next_cb
[2024-04-18T08:45:17.287Z]     return self._client.documents.search_post(search_request=self._initial_query.request, **self._kwargs)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
[2024-04-18T08:45:17.287Z]     return func(*args, **kwargs)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/search/documents/_generated/operations/_documents_operations.py", line 777, in search_post
[2024-04-18T08:45:17.287Z]     pipeline_response: PipelineResponse = self._client._pipeline.run(  # pylint: disable=protected-access
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 230, in run
[2024-04-18T08:45:17.287Z]     return first_node.send(pipeline_request)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.287Z]     response = self.next.send(request)
[2024-04-18T08:45:17.287Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   [Previous line repeated 2 more times]
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/policies/_redirect.py", line 197, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/policies/_retry.py", line 553, in send
[2024-04-18T08:45:17.288Z]     raise err
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/policies/_retry.py", line 531, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 86, in send
[2024-04-18T08:45:17.288Z]     response = self.next.send(request)
[2024-04-18T08:45:17.288Z]   [Previous line repeated 2 more times]
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/_base.py", line 119, in send
[2024-04-18T08:45:17.288Z]     self._sender.send(request.http_request, **request.context.options),
[2024-04-18T08:45:17.288Z]   File "/home/juanjo/juanjo_repos/functions/.venv/lib/python3.10/site-packages/azure/core/pipeline/transport/_requests_basic.py", line 386, in send
[2024-04-18T08:45:17.288Z]     raise error
[2024-04-18T08:45:17.288Z] .


already commented this on the discord this morning

**Comment by SimplyJuanjo** [comment]
thanks for ur response @HARISHKUMAR1112001 

it was really weird because both the AzureOpenAI and the SearchClient were working alone without DSPy

but

debugging with a custom AzureRM class i think i found the error

it was related to the "search_service_name=search_endpoint" arg

I was passing the search_endpoint as a complete url:

search_endpoint = "https://f29search.search.windows.net";

[2024-04-19T12:36:57.321Z] Azure Search Request
[2024-04-19T12:36:57.321Z] <SearchClient [endpoint='https://https://f29search.search.windows.net.search.windows.net', index='sharepoint-index']>

tried then without "https://"

search_endpoint = "https://f29search.search.windows.net";

[2024-04-19T12:41:45.467Z] Azure Search Request
[2024-04-19T12:41:45.469Z] <SearchClient [endpoint='https://f29search.search.windows.net.search.windows.net', index='sharepoint-index']>

but should have passed the service name only:

search_endpoint = "f29search"; # Add your Azure AI Search endpoint here

now it worked!

sorry for bothering, was my bad!

**Pull request merged**